programs/rodeos/streams/rabbitmq.hpp

```diff
@@ -21,41 +21,46 @@ class rabbitmq : public stream_handler {
    const std::string                    queue_name_;
    const std::vector<eosio::name>       routes_;
    const boost::filesystem::path        unconfirmed_path_;
-   std::shared_ptr<rabbitmq_handler>    handler_;
-   std::shared_ptr<AMQP::TcpConnection> connection_;
-   std::shared_ptr<AMQP::TcpChannel>    channel_;
    // capture all messages per block and send as one amqp transaction
    std::deque<std::pair<std::string, std::vector<char>>> queue_;
 
  public:
-   rabbitmq(boost::asio::io_service& io_service, std::vector<eosio::name> routes, const AMQP::Address& address, bool publish_immediately,
+   rabbitmq(std::vector<eosio::name> routes, const AMQP::Address& address, bool publish_immediately,
             std::string queue_name, const boost::filesystem::path& unconfirmed_path)
        : address_(address)
        , publish_immediately_(publish_immediately)
        , queue_name_( std::move( queue_name))
        , routes_( std::move( routes))
        , unconfirmed_path_( unconfirmed_path)
    {
-      ilog("Connecting to RabbitMQ address ${a} - Queue: ${q}...", ("a", std::string(address))( "q", queue_name_));
-
-      init(io_service, address);
-
-      declare_queue();
+      ilog("Connecting to RabbitMQ address ${a} - Queue: ${q}...", ("a", address)( "q", queue_name_));
+      bool error = false;
+      eosio::amqp declare_queue( address_, queue_name_, [&error](const std::string& err){
+         elog("AMQP Queue error: ${e}", ("e", err));
+         appbase::app().quit();
+         error = true;
+      } );
+      if( error ) return;
+      amqp_publisher_ = std::make_shared<eosio::reliable_amqp_publisher>(address_, "", "", unconfirmed_path_);
    }
 
-   rabbitmq(boost::asio::io_service& io_service, std::vector<eosio::name> routes, const AMQP::Address& address, bool publish_immediately,
+   rabbitmq(std::vector<eosio::name> routes, const AMQP::Address& address, bool publish_immediately,
             std::string exchange_name, std::string exchange_type, const boost::filesystem::path& unconfirmed_path)
        : address_(address)
        , publish_immediately_(publish_immediately)
        , exchange_name_( std::move( exchange_name))
        , routes_( std::move( routes))
        , unconfirmed_path_( unconfirmed_path)
    {
-      ilog("Connecting to RabbitMQ address ${a} - Exchange: ${e}...", ("a", std::string(address))( "e", exchange_name_));
-
-      init(io_service, address);
-
-      declare_exchange(exchange_type);
+      ilog("Connecting to RabbitMQ address ${a} - Exchange: ${e}...", ("a", address)( "e", exchange_name_));
+      bool error = false;
+      eosio::amqp declare_exchange( address_, exchange_name_, exchange_type, [&error](const std::string& err){
+         elog("AMQP Exchange error: ${e}", ("e", err));
+         appbase::app().quit();
+         error = true;
```

Comment:
data race.  The error handler can be called concurrently from both this thread and the amqp thread if `amqp::wait` times out.

Reply:
I first made all provided on_error provided callback functions thread-safe, but in the end I needed to use a mutex for `amqp::stop()` so just used same mutex to protect calls to on_error.